src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,312 @@
+            _collectionError = collectionError;
+
+            _listener = new MeterListener()
+            {
+                InstrumentPublished = (instrument, listener) =>
+                {
+                    _instrumentPublished(instrument);
+                    InstrumentState? state = GetInstrumentState(instrument);
+                    if (state != null)
+                    {
+                        _beginInstrumentMeasurements(instrument);
+                        listener.EnableMeasurementEvents(instrument, state);
+                    }
+                },
+                MeasurementsCompleted = (instrument, cookie) =>
+                {
+                    _endInstrumentMeasurements(instrument);
+                    RemoveInstrumentState(instrument, (InstrumentState)cookie!);
+                }
+            };
+            _listener.SetMeasurementEventCallback<double>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<float>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<long>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<int>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<short>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<byte>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<decimal>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+        }
+
+
+
+        public AggregationManager Include(string meterName)
+        {
+            Include(i => i.Meter.Name == meterName);
+            return this;
+        }
+
+        public AggregationManager Include(string meterName, string instrumentName)
+        {
+            Include(i => i.Meter.Name == meterName && i.Name == instrumentName);
+            return this;
+        }
+
+        public AggregationManager Include(Predicate<Instrument> instrumentFilter)
+        {
+            lock (this)
+            {
+                _instrumentConfigFuncs.Add(instrumentFilter);
+            }
+            return this;
+        }
+
+        public AggregationManager SetCollectionPeriod(TimeSpan collectionPeriod)
+        {
+            lock (this)
+            {
+                _collectionPeriod = collectionPeriod;
+            }
+            return this;
+        }
+
+        public void Start()
+        {
+            // if already started or already stopped we can't be started again
+            if (_collectThread != null || _cts.IsCancellationRequested)
+            {
+                // correct usage from internal code should never get here
+                throw new InvalidOperationException("Start can only be called once");
```

Comment:
>Start [](http://example.com/codeflow?start=53&length=5)

should we add these strings to the resources file?

Reply:
I'll follow whatever path you think is best. 
My logic for not adding them thus far is that a user should never be able to observe this exception unless our code has a bug in it. Would you like me to add them?

Reply:
Who will observe this exception message? is it only helpful for debugging? if so, then no need to store them in the resources but at least have some comments why we are doing that.

Reply:
Yeah, unless there are bugs in the implementation an end user should never see the exception. They are there just as defensive coding. In particular the ones that are checking the interval timing ensure we don't get into a tight infinite loop similar to a servicing issue we recently had to fix for EventCounter. I'll add comments.